Adds Jest and JSDoc, removes JQuery from `app.ts`

S0AndS0 opened the pull request:
**`.ci-jest/app.test.js` and `.ci-jest/view/audio.test.js`**

- adds files to organize testing project via Jest


**`README.md`**

- adds attribution section


**`package-lock.json` and `package.json`**

- adds `Jest` to development dependencies, configurations, and run 
scripts

- adds `JSDoc` to development dependencies


**`public/app.js` and `public/app.js.map`**

- re-builds files from TypeScript changes within `ts/app.js`


**`public/index.html`**

- updates renamed function and adds parameters

- corrects closure of `html` and `body` elements


**`ts/app.ts`**

- renames and splits `findMinAndMaxValues()` function

  - `getMinMaxValuesFrom(input)` reads from string like input
  - `setMinAndMaxValuesFrom(...html_ids)` reads and mutates HTML 
elements

- replaces JQuery code with strongly typed TypeScript

- replaces hard-coded element IDs with parameters for some functions

- adds notes and comments about some changes

- adds initial JSDoc strings


**`tsconfig.json`**

- adds module resolution config

- corrects indentation

> Thanks be to 
https://github.com/facebook/jest/issues/8218#issuecomment-486464992

oferb commented on ts/app.ts:
```diff
@@ -1,93 +1,141 @@
+
+/**
+ *
+ * @param {string} input -
```
`input -` - did you want to add some description after the `-`?

S0AndS0 replied:
:facepunch: Wups this one slipped past me; yeah I do want to add a description and will within the next PR.

oferb replied:
How will you remember? :)
Or do you mean within the next commit?

oferb commented on ts/app.ts:
```diff
@@ -1,93 +1,141 @@
+ * @param {string} max_html_id - HTML ID to set `.value` to max value
+ */
+function setMinAndMaxValuesFrom(
+  values_html_id: string = 'dataInput',
```
We use lowerCamelCase for variables, so `values_html_id` would be `valuesHtmlId`.

S0AndS0 replied:
Hmm, this is certainly a good reason to have some style guidelines.

Minimum I think we should have how to differentiate between...

- functions

- classes

- class methods

- variables of various scopes

... documented somewhere so that the code base remains consistent.

oferb replied:
We use this: https://google.github.io/styleguide/jsguide.html
Probably should add it to the contributing guidelines.

S0AndS0 replied:
Thank you! I'll be certain to bookmark it.

After quick review it seems as though functions, methods, and local variables all use the same styling... personally I find it easier to differentiate when there are differences, eg...

- `class` => `SomeClass`

- `method` => `aMethod`

- `function` => `Something_Functional`

- `variable` (local) => `internal_variable`

- `variable` (global) => `SCRIPT_VARIABLE`

... but I'll do my best to adhere to this project's current styling.

oferb replied:
Ok. Here's a quick link to the relevant section:
https://google.github.io/styleguide/jsguide.html#naming-rules-by-identifier-type

oferb commented on ts/app.ts:
```diff
@@ -1,93 +1,141 @@
+ * @link: https://stackoverflow.com/questions/1669190/#8986992
+ */
+function getMinMaxValuesFrom(input: string) {
+  // TODO: Add a method to parse the input data to a array of arrays for example,
```
I agree. In general, we should parse user data once into a data structure that's convenient for us, and then operate on that.

oferb commented on ts/app.ts:
```diff
@@ -1,93 +1,141 @@
+ *
+ * @param {string} input -
+ * @note: StackOverflow discussions may be worth review on why `forEach`
+ * @question: find-the-min-max-element-of-an-array-in-javascript
```
What is `@question`?

S0AndS0 replied:
`@question` would be the question's title, generally I try to provide enough hints for accessing linked to information even if a search engine is used instead of the direct link.

Also if I remember correctly JSDoc, when one get's fancy with it, allows for customizing how documentation is generated from various tags.

oferb replied:
So I think this should go inside the method's code, as code-level documentation, because it explains the implementation.

S0AndS0 replied:
Commit `c7afd58` from Fork if _squashed_ will remove notes. In the future I'll convert `for` loops into `forEach`, where applicable, because it'll prevent future issues if there's a large amount of data to loop over.

oferb replied:
We don't have lots of data (and probably won't in the future either), so I think readability has a stronger weight.
If `forEach` is more readable then yah - go for it.
Otherwise, it I wouldn't bother.

oferb commented on ts/app.ts:
```diff
@@ -1,93 +1,141 @@
+ * @param {string} input -
+ * @note: StackOverflow discussions may be worth review on why `forEach`
+ * @question: find-the-min-max-element-of-an-array-in-javascript
+ * @link: https://stackoverflow.com/questions/1669190/#8986992
```
Is the @note referring to the @link or are these separate things?

S0AndS0 replied:
Yeah the `@note` is related to the `@question` and `@link` lines, things were split such that lines didn't become to wide.

oferb commented on ts/app.ts:
```diff
@@ -1,93 +1,141 @@
+/**
+ *
+ * @param {string} input -
+ * @note: StackOverflow discussions may be worth review on why `forEach`
```
About the `@note` - I think if it refers to the forEach, it should probably be right above the forEach code. WDYT?

S0AndS0 replied:
You're probably correct, this is more of a notice about `forEach` use vs. other `for` loops in general.

oferb replied:
Still I think it should be closer to the actual code.
The function comment should detail mostly input and output, and not implementation details, unless relevant to the caller.

S0AndS0 replied:
I'll try to remember to place comments closer to code in the future.

oferb replied:
Thanks!

S0AndS0 commented on .ci-jest/view/audio.test.js:
```diff
@@ -0,0 +1,24 @@
+require('../../public/view/audio.js');
+
+function test_getCellMaxDistance() {
+  test('Does 1 === 1', () => {
```
I'll be converting JQuery within `ts/view/audio.ts` to TypeScript while adding tests within `.ci-jest/view/audio.test.js`. so at this point it's a placeholder to tell me where to expand upon in the future.

oferb replied:
Cool. Can you document that it's a placeholder?

S0AndS0 replied:
Sure commit hash `1a14a03` may be _squashed_ in if ya like

S0AndS0 commented on .ci-jest/app.test.js:
```diff
@@ -0,0 +1,32 @@
+    const select_node = document.createElement('select');
+    select_node.id = "ttsVoice";
+
+    expect(ttsVoiceSelect).toStrictEqual(select_node);
```
At the moment it's testing what happens when there are no languages supported, as well as providing an example of how to preform these sorts of tests with Jest, once I've sorted out how to mock browser features more I'll be adding further tests.

oferb replied:
Cool

S0AndS0 commented on package.json:
```diff
@@ -3,14 +3,49 @@
+      "node_modules/",
+      "public/jquery"
+    ],
+    "coverageThreshold": {
```
`coverageThreshold` would be the percentage of lines covered by tests, with current settings by default things require _`100`%_ if not otherwise defined. By setting files that I'm still sorting out tests for down to `0` this tells Jest to only warn of coverage gaps instead of flooding the terminal with error messages.

My goal is full coverage, however, this may not be feasible which is why I've setting blocks for defaults as well as for specific portions of this project.

oferb replied:
Cool. So default 100 is aspirational :-))

S0AndS0 replied:
Yep 100% is always the goal, and having it set as the default ensures that Jest will throw errors on new files if customizations are not defined.

If we start utilizing Travis-CI or GitHub Actions in the future this will mitigate chances of merging untested code into the main project too.

oferb replied:
Got it :)
Yah some CI would be nice. Could also be CircleCI.
I don't mind - we just need to run tests and compile the ts and see that the js is the same.

S0AndS0 commented on ts/app.ts:
```diff
@@ -1,93 +1,141 @@
+  const maxValue: string = (<HTMLInputElement> document.getElementById('maxValue')).value;
+  const instrumentType: string = (<HTMLInputElement> document.getElementById('instrumentType')).value;
+
+  // const ttsVoice: HTMLSelectElement = (<HTMLSelectElement> document.getElementById('ttsVoice'));
```
@oferb, tagging another GitHub account often requires an `@` prefix, eg...

@mohammad-suliman can ya let us know if the `ttsVoice` variable is safe to comment out?

mohammad-suliman replied:
Hello @S0AndS0 
First, let me thank you for your great work and the efforts and the devotion for the project!! Me and @oferb appreciate that so much...
Yes, I think those lines can be removed, the code at line 29 sub seeds them.

S0AndS0 replied:
@mohammad-suliman, you're (both) very welcome! It's an enjoyable experience to collaborate with a talented team.

I've removed the commented code within the latest fork push, and aside from one questionable line (the element ID `updateButton` doesn't seem to exist within any HTML document), this PR _should_ be ready to be squashed and merged in.

S0AndS0 commented on ts/app.ts:
```diff
@@ -1,93 +1,141 @@
+    /**
+     * @note: the following element does **not** seem to exist
+     */
+    (<HTMLInputElement> document.getElementById('updateButton')).disabled = false;
```
@mohammad-suliman I couldn't find any HTML elements with the ID of `updateButton` within this project, is the JavaScript code still needed or is it okay to remove this reference?

S0AndS0 commented:
Commits `c7afd5`, `afc9631`, `1a14a03` and `8c6a36d` may be squash merged in, and _should_ address most of the requested changes.

I'm not certain that GitHub has such features, so via Git CLI it may be as easy as...


```Bash
cd ~/git/hub/hasadna/sensory-interface

git remote add fork git@github.com:S0AndS0/sensory-interface.git

git pull fork

git checkout master

git merge --squash fork/master

git commit -m 'Merges `S0AndS0/sensory-interface` into `hasadna/sensory-interface`'

git push origin master
```

oferb commented:
GitHub has squash commits, so we should be able to squash them :)
Thanks for the contrib!!

S0AndS0 commented:
Note the merge conflict within `public/index.html` can be resolved by accepting purposed changes, eg...

```HTML
    <textarea id="dataInput"
<<<<<<< master
      onblur="setMinAndMaxValuesFrom(this.id, 'minValue', 'maxValue')">1272	1289	1291	1291	1306	1309	1298	1297	1296	1309	1333	1319	1312	1301	1300	1293	1305	1313	1312	1304	1288</textarea>
=======
      onblur="findMinAndMaxValues()">1272	1289	1291	1291	1306	1309	1298	1297	1296	1309	1333	1319	1312	1301	1300	1293	1305	1313	1312	1304	1288</textarea>
>>>>>>> master
    <br>
```


... should be changed to...


```HTML
    <textarea id="dataInput"
      onblur="setMinAndMaxValuesFrom(this.id, 'minValue', 'maxValue')">1272	1289	1291	1291	1306	1309	1298	1297	1296	1309	1333	1319	1312	1301	1300	1293	1305	1313	1312	1304	1288</textarea>
    <br>
```


I think the reason for Git barfing on this file is that I've added parameters and renamed the function to be more expressive of what actions are being preformed.
